rust/kernel/platdev.rs

```diff
@@ -15,15 +15,15 @@ use crate::{
     types::PointerWrapper,
 };
 use alloc::boxed::Box;
-use core::{marker::PhantomPinned, pin::Pin};
 
 /// A registration of a platform device.
-#[derive(Default)]
+///
+/// # Invariants
+///
+/// We never move out of `pdrv`'s `Box`.
 pub struct Registration {
-    registered: bool,
     of_table: Option<*const c_types::c_void>,
-    pdrv: bindings::platform_driver,
-    _pin: PhantomPinned,
+    pdrv: Box<bindings::platform_driver>,
```

Comment:
Would the code be much more complicated if you had `Pin<Box<...>>` here?

My impression is that it wouldn't be too different (after the simplification you're applying), and the compiler would give you a little bit of help in enforcing the invariant.

Reply:
Thanks for the review! I'm glad you are asking this question. I can use it as a vehicle to check and update my understanding of the whole `Pin`ning concept.

As far as I know, using `Pin<Box<bindings::platform_driver>>` on its own here won't change a thing. It won't prevent the `platform_driver` from moving out of the `Box`. Why? Because `platform_driver: Unpin` by default. So we can always call [`Pin::into_inner()`](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.into_inner) safely, and the inner object will simply move out without complaints.

If we really want to prevent moving out, we need to mark `platform_driver` as `!Unpin`:
```rust
impl !Unpin for bindings::platform_driver {}
```
But Rust doesn't currently allow that!
```
error[E0658]: negative trait bounds are not yet fully implemented; use marker types for now
  --> rust/kernel/platdev.rs:20:6
   |
20 | impl !Unpin for bindings::platform_driver {}
```
So we'd have to use `PhantomPinned`, which in turn requires us to use some wrapper type:
```rust
struct pinned_platform_driver {
    inner: bindings::platform_driver,
    _pin: PhantomPinned,
}
// can't move out of Pin<Box<pinned_platform_driver>>, yay !
```
But now we are storing a different structure, one we can't get an inner mutable reference to without using `unsafe` blocks. Yet the kernel's `platform_register()`/`platform_unregister()` require mutable references !

During registration, it's not so bad, because  we can register first, then pin. Which kind of defeats the reason for using `Pin<>` in the first place..  But the unregistration function now becomes problematic: it needs a mutable reference to a `bindings::platform_driver`, but that structure is wrapped in a `!Unpin` outer wrapper, so we need a bunch of `unsafe` acrobatics there to get the reference.

As far as I can tell, it all becomes a huge unreadable, unsafe mess. IMHO using a "do not move out" Invariant works pretty well. But I am open to suggestions.

Reply:
It would not be an invariant, no? i.e. an invariant is something that is guaranteed (as long as one does not break preconditions on e.g. constructors).

Reply:
It would be an invariant, but not one the compiler will help enforcing.

Reply:
If want to avoid using `Pin<>` here (and I think that's probably the way to go, subject to further feedback), then we'll need a way to document / enforce that no code inside `platdev::Registration` moves `pdrv` out of its `Box`. We should not need to enforce this externally, as `pdrv` is a private member field, and is never exposed externally. Except through a kernel function, but that doesn't count.

What is the best way to accomplish that? Using an invariant? Or some other way?

Reply:
@bjorn3 I am talking about the "do not move out" comment -- if it is written like that, it isn't an invariant.

Reply:
@TheSven73 If we only document it, then there would be ways to trigger UB without writing `unsafe` (in the implementation), even if the abstraction happens to be sound for clients. The problem I see with that is that we would be opening the "gates" to do the same trick everywhere else in the `kernel` crate for other things.

Reply:
@ojeda are you saying that there's no "proper" way to write this as a type invariant on `struct Registration`? Is this because type invariants are properties that must be maintained, and "do not move out" is more like a prohibition?

Trying to understand what people here see as the issue.

Reply:
The safety boundary in rust is a module. Take for example `Vec`. It has the invariant that the inner pointer is valid and the length and capacity are correct. However any safe code inside the `alloc::vec` module can write to those fields, so you have to audit the whole module if you want to check that any unsafe code depending on invariants of `Vec` are sound.

Reply:
Regarding the comment, sorry for the confusion: I thought you wanted to replace the comment you already have in the commit with the "do not move" one (which indeed looks like a prohibition for users, which is why I said it did not seem like an invariant to me). So that is fine -- also something like `pdrv is pinned, i.e. is never moved out` would work too.

Concerning the invariant vs. `Pin`: the advantage of using `Pin` is that, even in the implementation of the abstraction, it would ensure we cannot move things by mistake. In other words, if we only say the member intended to be pinned, we can still of course have a sound abstraction for users, but then we lose the "protection" for ourselves (e.g. making a mistake would be easier when modifying the implementation of the abstraction in the future).

Going back to what you said, if I understood you correctly, what you are saying is that using `Pin` properly here (i.e. with a wrapper type) requires too much extra complexity without buying us much, which is fair. My "worry" is that we start avoiding to use `Pin` and we lose that "layer of defense" (plus we lose the experience of how to use it, we would not care too much about trying to improve its ergonomics, etc.). But perhaps that is the way to go until there is a better way to express pinned data in Rust...

Reply:
> The safety boundary in rust is a module. Take for example `Vec`. It has the invariant that the inner pointer is valid and the length and capacity are correct. However any safe code inside the `alloc::vec` module can write to those fields, so you have to audit the whole module if you want to check that any unsafe code depending on invariants of `Vec` are sound.

Yes, of course. My point was that by leaving the pinning as a comment, then it is easier to make a mistake in the implementation. I edited the comment to add "more" to "ways to trigger UB" to try to be more clear.

Reply:
No worries.

> Going back to what you said, if I understood you correctly, what you are saying is that using `Pin` properly here (i.e. with a wrapper type) requires too much extra complexity without buying us much, which is fair.

I'm arguing that it's even worse than that: it requires a lot of extra complexity, **and** adds unsafe blocks, without buying us much. The "do not move out of the box" technique in this PR doesn't need any unsafe blocks at all.

As @bjorn3 is saying, it's perfectly possible that someone else from inside the crate somehow "steals" `pdev` out of its box. That's of course also possible when using `Pin<Box<>>` - except that the thief must now use an `unsafe` block.

So do people here prefer `Pin<>` because it requires thieves to use an unsafe block? What is the advantage?

> My "worry" is that we start avoiding to use `Pin` and we lose that "layer of defense" (plus we lose the experience of how to use it, we would not care too much about trying to improve its ergonomics, etc.). But perhaps that is the way to go until there is a better way to express pinned data in Rust...

I'd like to understand what is meant by "that layer of defense".

Reply:
> As @bjorn3 is saying, it's perfectly possible that someone else from inside the crate somehow "steals" `pdev` out of its box. That's of course also possible when using `Pin<Box<>>` - except that the thief must now use an `unsafe` block.

Yeah, the "layer of defense" would be that now making that mistake is harder since it requires `unsafe`, i.e. they would not forget about the invariant since it is now encoded in the type.

It is like any other use of the type system to encode invariants: it helps avoiding mistakes.

But in this case, given we even need to introduce extra types (`pinned_platform_driver`) to wrap things to make them actually pinned, etc. etc., then it is likely not worth it. So I am OK with the approach you take here. I would nevertheless add a comment explaining this, e.g. something like "pinning this via the type system requires X, Y, Z and thus we decide to keep code simple -- the abstraction is still sound for users since the member is not exposed etc etc".

Reply:
> So I am OK with the approach you take here. I would nevertheless add a comment explaining this, e.g. something like "pinning this via the type system requires X, Y, Z and thus we decide to keep code simple -- the abstraction is still sound for users since the member is not exposed etc etc".

Sounds like a plan !

That said, I am still unbelievably confused by the use of `Pin`ning in Rust-for-Linux. All over the codebase, it's assumed that if `T: !Unpin`, then data inside `Pin<Box<T>>` is safe to pass to the kernel, because it can never move. At least, not without unsafe.

But I am not sure that's actually true. Consider this: ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e639814b2b5b557ac3c4aec8e9b7e5a9))
```rust
use std::pin::Pin;
use std::marker::PhantomPinned;
use std::boxed::Box;

#[repr(C)]
#[derive(Default)]
struct platform_driver {
    f: Option<unsafe extern "C" fn()>,
    _pin: PhantomPinned,
}

extern "C" fn callback_1() {
    println!("callback 1");
}

extern "C" fn callback_2() {
    println!("callback 2");
}

fn main() {
    let mut pd1 = platform_driver::default();
    pd1.f = Some(callback_1);
    let mut pd1: Pin<Box<platform_driver>> = Box::pin(pd1);
    // Pin<Box<platform_driver>> where platform_driver: !Unpin is allegedly
    // safe to be passed to C. Because it can never move out using only safe code.
    println!("pd1 pointer address: {:?}", &pd1 as *const _);
    unsafe { pd1.f.unwrap()(); }
    // OH REALLY ?
    let mut pd2 = platform_driver::default();
    pd2.f = Some(callback_2);
    pd1.set(pd2);
    println!("pd1 pointer address: {:?}", &pd1 as *const _);
    unsafe { pd1.f.unwrap()(); }
    // pd1's underlying memory has been modified, even while it's marked
    // !Unpin **and** located behind Pin<>.
    // What would have happened had a pointer to this struct been in use with the kernel?
}
```
Running this gives us:
```
pd1 pointer address: 0x7ffca1641368
callback 1
pd1 pointer address: 0x7ffca1641368
callback 2
```

Reply:
In your example, the memory location has not moved [*], so if someone has a pointer to there, it is still valid. Of course, the data at the address can be modified -- so I am confused about what you mean.

If what you mean is that the function pointer address has changed, that can happen too since the example pins the outer struct, not the function pointer.

[*] Note that those addresses in your example point to the stack, i.e. to the box, not to the pinned memory. To print the address of the actual data, the easiest is to use `{:p}` directly on the box; and you can also use it for the address of variables, e.g.

```rust
println!("address of the data: {:p}", pd1);
println!("address of the  box: {:p}", &pd1);
```

Reply:
Thanks for the tip! I'm sorry that my confusion is so confusing :) I'd like to ask a few questions, to update my internal model.

When we register a miscdev for example, we pass a pointer to a `bindings::miscdevice` to the kernel on registration. I'm assuming we need two guarantees:
- the struct's memory area will continue to exist until we unregister the `miscdevice` and
- the struct's memory area won't be overwritten until after we unregister the `miscdevice`

The easiest way is probably like this: (a)
```rust
struct Registration {
    mdev: Box<bindings::miscdevice>,
    // etc
}
impl Registration {
    fn new() -> Self {
        let mut mdev = Box::new(default());
        unsafe { bindings::misc_register(&mut *mdev) };
        Self { mdev }
    }
}
```
Yet the existing codebase does the following: (b)
```rust
struct Registration {
    mdev: bindings::miscdevice,
    _pin: PhantomPinned,
}
impl Registration {
    fn new() -> Pin<Box<Self>> {
        let mut r = Pin::from(Box::new(Self::default()));
        let mut this = unsafe { /* r.get_mut() and get_unchecked_mut() gymnastics */ };
        unsafe { bindings::misc_register(&mut this.mdev) };
        r
    }
}
```
So my question is: why does the current codebase choose (b) over (a)? What are the advantages of (b) over (a)?
According to the comments in the code,
```rust
    /// It must be pinned because the memory block that represents the registration is
    /// self-referential.
```
but I am not sure that's a sufficient justification, because (a) is not `Pin`ned, yet also allows self-referential structures.

Reply:
Sounds correct to me. If one keeps the things that need to be pinned inside a `Box`, and the box remains private (i.e. others cannot mess with it), and the implementation of the abstraction takes care of never moving it, then it is OK. It is also OK to `Pin` the private `Box` to ensure it is not moved by mistake by the implementation.

However, IIRC there was another reason to have (in this case) `Registration` pinned, rather than leave it a detail of the implementation. @wedsonaf?

Reply:
Option (b) isn't really what we have in the current codebase. What we have is (simplified):
```rust
impl Registration {
    fn new() -> Self {
        [...]
    }
}
```

Then as a helper we have:

```rust
    fn new_pinned() -> Result<Pin<Box<Self>>> {
        [...]
    } 
```

And the reason is that we want to have a zero-cost abstraction. If we forced everyone to use `new_pinned` or option (a) above, we would require everyone to have an extra allocation. Here's an example (C Binder) on how miscdev is embedded in some outer structure: https://elixir.bootlin.com/linux/latest/source/drivers/android/binder_internal.h#L36 -- it would require an extra allocation if we were to require `miscdev` to be always boxed; such a  solution would _preclude_ the zero-cost case, but the current implementation allows both: if you want zero-cost, you embed `Registration` in your outer structure and ensure that it's pinned (by pinning the outer structure and ensuring that the [projection is also pinned](https://doc.rust-lang.org/std/pin/index.html#pinning-is-structural-for-field)), but if you don't care about zero-cost, just use `new_pinned()` and pay the cost of the extra allocation.

Regarding Sven's example, that is perfectly fine: note that `pd1.set(pd2)` causes the destructor of whatever was pinned in `pd1` to be called before it is overwritten by `pd2`, that is, if what was initially pinned was a `Registration`, its destructor would unregister it first, then it would be overwritten with the new one.

Reply:
@wedsonaf and @ojeda, thank you so much for your patience and willingness to explain, I really appreciate it !! Let’s see if I understand the underlying issues better now. I encourage you to update my understanding if something doesn’t sound quite right.

When we create an abstraction that uses a kernel C structure in Rust, e.g. via `misc_register()` or `__mutex_init()`, there are a couple of “gotchas”:
1. We must not allow our users to modify the structure, other than through the proper C APIs and/or rules.
1. We must not allow our users to move the kernel C structure around in memory.

(1)  is easy. Users from outside the kernel crate can’t see private fields, so we can always make the C structure private, and ensure our public functions use it as prescribed in the kernel.

(2) is trickier! If the C structure is just a straight-forward field inside our Rust struct, then our users could move it around in memory, which breaks the C API. There are two solutions to this problem:

1. Store the C structure in a private, `Box`ed field of the Rust struct. Our users can’t touch it because it’s private. They also can’t move it around - moving the Rust structure in memory is fine, because it’s just holding a pointer to the `Box`ed C struct.
1. Store the C structure inside the Rust structure in a straightforward way. And prevent our users from moving it around using various techniques. Such as `# Safety` requirements, or method functions with a `self` type that’s `Pin<&Self>` or `Pin<&mut Self>`. All of these require our users to use unsafe blocks and make careful reasonings about pinning and safety.

All are partially flawed. Solution (1) leads to multiple allocations and indirections, which is a non-zero cost abstraction - hey that’s not what Rust is supposed to be about! Solution (2) is true zero-cost, yet requires users to write tricky unsafe code just to use what are essentially very simple abstractions. They have a good chance of getting it wrong, and go down in flames - hey that’s also not what Rust is all about!

I am not sure what the group consensus is here. Are the unsafe blocks and tricky pin projections in driver code not seen as a problem?

For `platdev` I will go with solution (1) for now, which is easy to use for driver writers, yet leads to multiple allocations and indirections. For a stone-cold path such as registering a platform driver (not a device), that shouldn't be a problem.

Reply:
I think your understanding is correct!

I think it is reasonable to favor the easiest solution for cold paths, although people may still complain because, in the case of allocations, they may lead to other issues outside the cold path (say, memory fragmentation with SLOB).

Then there is the problem of having to return more `Result`s. And, in general, the overall feeling that the language "forces" us to do allocations just to have a nicer API.

I don't think what way would be best for us, but I tend to think that medium-/long-term we should invest in finding out the best ergonomic solution while keeping the zero-cost (via the `pin-*`-like crates, a new language feature, or something else).